soroban-sdk/src/contract_event.rs

```diff
@@ -24,43 +32,91 @@ use crate::{env::internal, Env, IntoVal, RawVal, Vec};
 /// # #[cfg(feature = "testutils")]
 /// # fn main() {
 /// #     let env = Env::default();
-/// #     let contract_id = FixedBinary::from_array(&env, [0; 32]);
+/// #     let contract_id = BytesN::from_array(&env, [0; 32]);
 /// #     env.register_contract(&contract_id, Contract);
 /// #     f::invoke(&env, &contract_id);
 /// # }
 /// # #[cfg(not(feature = "testutils"))]
 /// # fn main() { }
 /// ```
 #[derive(Clone)]
-pub struct ContractEvent(Env);
+pub struct Events(Env);
 
-impl Debug for ContractEvent {
+impl Debug for Events {
     fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-        write!(f, "ContractEvent")
+        write!(f, "Events")
+    }
+}
+
+pub trait IntoTopics {
+    fn into_topics(self, env: &Env) -> Object;
+}
+
+// 0 topics
+impl IntoTopics for () {
+    fn into_topics(self, env: &Env) -> Object {
+        Vec::<RawVal>::new(env).to_object()
     }
 }
 
-impl ContractEvent {
+macro_rules! impl_for_tuple {
```

Comment:
Nitpick: can you give this a name that indicates what is being `impl`ed?

I also think you should evaluate the impact of this approach on size of WASM binaries. One different approach that might or might not be more size efficient would be to write a private function that takes `Vec<EnvVal>` (or something like it) and contains all the logic then just provide public tuple wrappers to that function.

Reply:
Yeah, this is a great point! I will definitely check the sizes. This might also have a larger size than @leighmcculloch 's [approach](https://github.com/stellar/rs-soroban-sdk/pull/467) since it has extra topic validation logic.

Reply:
I think we can keep the validation logic in the host side. If the WASM side is just going to panic, the host can do that.

Reply:
This makes sense, I've removed the validation logic.